File: core/FindFirstAndLastPositionOfElementInSortedArray/FindFirstAndLastPositionOfElementInSortedArray.py
from typing import List


def searchRange(nums: List[int], target: int) -> List[int]:
    res = [-1, -1]

    def binary_search(left, right):
        if left > right:
            return
        mid = (left + right) // 2
        if nums[mid] < target:
            binary_search(mid + 1, right)
        elif nums[mid] > target:
            binary_search(left, mid - 1)
        else:
            res[0] = mid if res[0] == -1 else min(mid, res[0])
            res[1] = mid if res[1] == -1 else max(mid, res[1])
            binary_search(left, mid - 1)
            binary_search(mid + 1, right)

    binary_search(0, len(nums) - 1)
    return res

File: core/FindFirstAndLastPositionOfElementInSortedArray/FindFirstAndLastPositionOfElementInSortedArray.java
package com.dbc;

public class FindFirstAndLastPositionOfElementInSortedArray {
    private void binarySearch(int left, int right, int[] nums, int target, int[] res) {
        if (left > right) return;
        int mid = (left + right) / 2;
        if (nums[mid] > target) binarySearch(left, mid - 1, nums, target, res);
        else if (nums[mid] < target) binarySearch(mid + 1, right, nums, target, res);
        else {
            res[0] = res[0] == -1 ? mid : Math.min(res[0], mid);
            res[1] = res[1] == -1 ? mid : Math.max(res[1], mid);
            binarySearch(left, mid - 1, nums, target, res);
            binarySearch(mid + 1, right, nums, target, res);
        }
    }

    public int[] searchRange(int[] nums, int target) {
        int[] res = new int[]{-1, -1};
        binarySearch(0, nums.length - 1, nums, target, res);
        return res;
    }
}
